includes/class-plugin.php

```diff
@@ -984,17 +981,13 @@ public function do_admin_actions() {
                         $result = $wp_filesystem->delete( WP_CONTENT_DIR . '/object-cache.php' );
 
                         if ( $result ) {
-                            try {
-                                $predis = new Predis();
-                                $predis->flush();
-                            } catch ( Exception $exception ) {
-                                // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
-                                error_log( $exception );
-                            }
```

Comment:
Why don't we use this for flushing, but set the `read_timeout` _much_ higher in `Rhubarb\RedisCache\Predis`, so that Redis has enough time to complete the operation?

Reply:
I don't think the problem is the `read_timeout` on the flush, it's the subsequent errors on cache requests whilst the Redis server is flushing the cache. You also can't guess how long this process will take even if setting `read_timeout` would prevent the issue (which I don't think it will) because different cache setups will take different amounts of time to flush.

Reply:
Redis commands are blocking so the PHP-execution is halted until Redis completed the `FLUSHDB` operation.

After that PHP-execution will resume and subsequent Redis calls should not timeout (`read error on connection`). Unless that's what you're seeing (subsequent calls fail, not the flush itself)?

Reply:
It is my understanding that the subsequent calls are failing because the Redis server has not finished flushing.

Please note that using this patch we have not seen any errors in debug.log when flushing the cache so it would appear that the flush function of Redis does not block PHP.

If what you are saying were true then we would see errors no matter where the cached was flushed (at the point of the Settings API hook, the `shutdown` hook or any hook). So it is my understanding that it is not the flush execution that is causing the errors but the subsequent cache requests.

Reply:
Your understanding of how PHP and Redis operate (in this case) is incomplete. 

All Redis commands are atomic and blocking. PHP will wait until Redis responds or times-out.

These two proven facts are contrary to your statements. 

I think the next logical step is to prove whether the flush or a subsequent command fails. For that let's look at the exact PHP error logs when you encounter the "read error on connection" issue.

Reply:
My understanding aside, I’ll get the information you require and report back.

Reply:
I wonder if the solution here is to bump the read_timeout for your installation. And later introduce a new flush timeout that won’t be as harsh as the default 1.0 timeout.

Reply:
OK, so it looks like it _is_ the flush that's failing ...

```
[07-Oct-2023 07:02:29 UTC] Before flush cache ...
[07-Oct-2023 07:02:30 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:1647
Stack trace:
#0 /var/www/html/wp-content/object-cache.php(1647): Redis->flushDB()
#1 /var/www/html/wp-content/object-cache.php(149): WP_Object_Cache->flush()
#2 /var/www/html/wp-content/plugins/redis-cache/includes/class-plugin.php(925): wp_cache_flush()
#3 /var/www/html/wp-includes/class-wp-hook.php(310): Rhubarb\RedisCache\Plugin->do_admin_actions()
#4 /var/www/html/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters()
#5 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action()
#6 /var/www/html/wp-admin/admin.php(237): do_action()
#7 /var/www/html/wp-admin/options-general.php(10): require_once('/var/www/html/w...')
#8 {main}
[07-Oct-2023 07:02:58 UTC] Before flush cache ...
[07-Oct-2023 07:02:58 UTC] ... after flush cache
```

... but I'm struggling to test the "flush after disable" because as soon as the cache has been flushed I need to wait until the cache has built up again so I will test that again tomorrow. I have two sizeable sites that I can test this on.

It is the latter "flush after disable" where there were multiple errors which lead me to believe the subsequent cache requests were failing post flush but I guess this _could_ be a race condition.

If the solution here **is** to increase the flush timeout ... can this be done independently of the global `timeout` setting?

Also, I am assuming that `Predis()->flush()` and `wp_cache_flush()` use the same `timeout` setting?

Reply:
Sweet, okay. We can improve this.

`timeout` is only used when connecting to Redis, while `read_timeout` is used when reading/writing/flushing. The misleading naming is legacy naming from PhpRedis.

`Rhubarb\RedisCache\Predis::flush()` is its own dedicated flush class, while `wp_cache_flush()` uses what `WP_Object_Cache` provides.

I've stubbed out some ideas here: https://github.com/rhubarbgroup/redis-cache/pull/491

Reply:
In the meantime, you can bump `WP_REDIS_READ_TIMEOUT` to avoid the issue on your large sites.

Reply:
Sorry for my delay in replying, I think I need to wait a couple of weeks before testing the deactivate Redis errors (as there were definitely more than one). I will update in due course.

Reply:
No worries, we're quite busy over here as well, we'll try to get #491 out the door.

Reply:
OK.

Interestingly, we noticed that on all three sites that use this plugin, when a WordPress core update was applied (v6.3.1) the installation process was interrupted with the same timeout error and the next visit to the dashboard required a WordPress database update to be run (other than that the core update appears to have been applied successfully despite the interruption).

Is this connected? Is there a cache flush run when a core update is in progress? Or should Redis be disabled prior to a WordPress core update? In which case automatic core updates should be disabled?

Reply:
Yikes. WordPress core will call `wp_cache_flush()` during the update to avoid stale data.

Reply:
Thought so!

Reply:
Didn't quite get there in time before WordPress automatically updated to v6.3.2 this morning but this is what happened ... I think the upgrade went OK despite these errors though ....

> [13-Oct-2023 08:44:46 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:1647
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(1647): Redis->flushDB()
> #1 /var/www/html/wp-content/object-cache.php(149): WP_Object_Cache->flush()
> #2 /var/www/html/wp-admin/includes/update-core.php(1585): wp_cache_flush()
> #3 /var/www/html/wp-admin/includes/class-core-upgrader.php(174): update_core()
> #4 /var/www/html/wp-admin/update-core.php(895): Core_Upgrader->upgrade()
> #5 /var/www/html/wp-admin/update-core.php(1151): do_core_upgrade()
> #6 {main}
> [13-Oct-2023 08:44:46 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:46 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2059
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(2059): Redis->mget()
> #1 /var/www/html/wp-content/object-cache.php(208): WP_Object_Cache->get_multiple()
> #2 /var/www/html/wp-includes/functions.php(7055): wp_cache_get_multiple()
> #3 /var/www/html/wp-includes/post.php(7763): _get_non_cached_ids()
> #4 /var/www/html/wp-includes/class-wp-query.php(3317): _prime_post_caches()
> #5 /var/www/html/wp-includes/class-wp-query.php(3800): WP_Query->get_posts()
> #6 /var/www/html/wp-includes/post.php(2441): WP_Query->query()
> #7 /var/www/html/wp-content/plugins/redacted/lib/functions.php(2610): get_posts()
> #8 /var/www/html/wp-content/plugins/redacted/lib/shortcodes.php(462): redacted_functions::get_listings()
> #9 /var/www/html/wp-includes/shortcodes.php(395): redacted_shortcodes::listings_column_block()
> #10 [internal function]: do_shortcode_tag()
> #11 /var/www/html/wp-includes/shortcodes.php(235): preg_replace_callback()
> #12 /var/www/html/wp-content/plugins/redacted/lib/shortcodes.php(224): do_shortcode()
> #13 /var/www/html/wp-includes/shortcodes.php(395): redacted_shortcodes::listings_map_towns()
> #14 [internal function]: do_shortcode_tag()
> #15 /var/www/html/wp-includes/shortcodes.php(235): preg_replace_callback()
> #16 /var/www/html/wp-includes/block-template.php(230): do_shortcode()
> #17 /var/www/html/wp-includes/template-canvas.php(12): get_the_block_template_html()
> #18 /var/www/html/wp-includes/template-loader.php(106): include('/var/www/...')
> #19 /var/www/html/wp-blog-header.php(19): require_once('/var/www/...')
> #20 /var/www/html/index.php(17): require('/var/www/...')
> #21 {main}
> [13-Oct-2023 08:44:46 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:46 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:47 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:47 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:47 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:47 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:48 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-admin/admin-ajax.php(22): require_once('/var/www/...')
> #7 {main}
> [13-Oct-2023 08:44:48 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:49 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-admin/admin-ajax.php(22): require_once('/var/www/...')
> #7 {main}
> [13-Oct-2023 08:44:49 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-admin/admin-ajax.php(22): require_once('/var/www/...')
> #7 {main}
> [13-Oct-2023 08:44:50 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/...')
> #7 /var/www/html/index.php(17): require('/var/www/...')
> #8 {main}
> [13-Oct-2023 08:44:50 UTC] RedisException: socket error on read socket in /var/www/html/wp-content/object-cache.php:563
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(563): Redis->ping()
> #1 /var/www/html/wp-content/object-cache.php(256): WP_Object_Cache->__construct()
> #2 /var/www/html/wp-includes/load.php(851): wp_cache_init()
> #3 /var/www/html/wp-settings.php(131): wp_start_object_cache()
> #4 /var/www/html/wp-config.php(124): require_once('/var/www/...')
> #5 /var/www/html/wp-load.php(50): require_once('/var/www/...')
> #6 /var/www/html/wp-admin/admin-ajax.php(22): require_once('/var/www/...')
> #7 {main}

Reply:
> Sorry for my delay in replying, I think I need to wait a couple of weeks before testing the deactivate Redis errors (as there were definitely more than one). I will update in due course.

Hi again, OK, we now have the log data showing multiple errors when disabling Redis ... (please note that the errors persist **after** the disable Redis cache flush has finished running ...)

> [19-Oct-2023 08:52:06 UTC] Before disable cache flush  ...
> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(2161): Redis->set()
> #1 /var/www/html/wp-content/object-cache.php(294): WP_Object_Cache->set()
> #2 /var/www/html/wp-content/plugins/redacted/redacted.php(3106): wp_cache_set()
> #3 /var/www/html/wp-includes/class-wp-hook.php(310): wordpress_maintenance_class::pre_option()
> #4 /var/www/html/wp-includes/plugin.php(205): WP_Hook->apply_filters()
> #5 /var/www/html/wp-includes/option.php(150): apply_filters()
> #6 /var/www/html/wp-includes/theme.php(183): get_option()
> #7 /var/www/html/wp-includes/theme.php(194): get_stylesheet()
> #8 /var/www/html/wp-includes/block-template-utils.php(271): get_stylesheet_directory()
> #9 /var/www/html/wp-includes/block-template-utils.php(775): _get_block_template_file()
> #10 /var/www/html/wp-includes/block-template-utils.php(1005): _build_block_template_result_from_post()
> #11 /var/www/html/wp-includes/blocks/template-part.php(232): get_block_templates()
> #12 /var/www/html/wp-includes/blocks/template-part.php(272): build_template_part_block_instance_variations()
> #13 /var/www/html/wp-includes/blocks/template-part.php(285): build_template_part_block_variations()
> #14 /var/www/html/wp-includes/class-wp-hook.php(310): register_block_core_template_part()
> #15 /var/www/html/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters()
> #16 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action()
> #17 /var/www/html/wp-settings.php(632): do_action()
> #18 /var/www/html/wp-config.php(124): require_once('/var/www/redacted...')
> #19 /var/www/html/wp-load.php(50): require_once('/var/www/redacted...')
> #20 /var/www/html/wp-blog-header.php(13): require_once('/var/www/redacted...')
> #21 /var/www/html/index.php(17): require('/var/www/redacted...')
> #22 {main}
> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(2161): Redis->set()
> #1 /var/www/html/wp-content/object-cache.php(294): WP_Object_Cache->set()
> #2 /var/www/html/wp-content/plugins/redacted/redacted.php(3106): wp_cache_set()
> #3 /var/www/html/wp-includes/class-wp-hook.php(310): wordpress_maintenance_class::pre_option()
> #4 /var/www/html/wp-includes/plugin.php(205): WP_Hook->apply_filters()
> #5 /var/www/html/wp-includes/option.php(150): apply_filters()
> #6 /var/www/html/wp-includes/theme.php(672): get_option()
> #7 /var/www/html/wp-includes/theme.php(353): get_theme_root_uri()
> #8 /var/www/html/wp-content/themes/twentytwentytwo/inc/patterns/hidden-heading-and-bird.php(19): get_template_directory_uri()
> #9 /var/www/html/wp-content/themes/twentytwentytwo/inc/block-patterns.php(131): require('/var/www/redacted...')
> #10 /var/www/html/wp-includes/class-wp-hook.php(310): twentytwentytwo_register_block_patterns()
> #11 /var/www/html/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters()
> #12 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action()
> #13 /var/www/html/wp-settings.php(632): do_action()
> #14 /var/www/html/wp-config.php(124): require_once('/var/www/redacted...')
> #15 /var/www/html/wp-load.php(50): require_once('/var/www/redacted...')
> #16 /var/www/html/wp-blog-header.php(13): require_once('/var/www/redacted...')
> #17 /var/www/html/index.php(17): require('/var/www/redacted...')
> #18 {main}
> [19-Oct-2023 08:52:07 UTC] ... after disable cache flush
> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(2161): Redis->set()
> #1 /var/www/html/wp-content/object-cache.php(294): WP_Object_Cache->set()
> #2 /var/www/html/wp-content/plugins/redacted/redacted.php(3106): wp_cache_set()
> #3 /var/www/html/wp-includes/class-wp-hook.php(310): wordpress_maintenance_class::pre_option()
> #4 /var/www/html/wp-includes/plugin.php(205): WP_Hook->apply_filters()
> #5 /var/www/html/wp-includes/option.php(150): apply_filters()
> #6 /var/www/html/wp-includes/theme.php(183): get_option()
> #7 /var/www/html/wp-includes/theme.php(121): get_stylesheet()
> #8 /var/www/html/wp-includes/class-wp-theme-json-resolver.php(241): wp_get_theme()
> #9 /var/www/html/wp-includes/class-wp-theme-json-resolver.php(591): WP_Theme_JSON_Resolver::get_theme_data()
> #10 /var/www/html/wp-includes/script-loader.php(3249): WP_Theme_JSON_Resolver::get_merged_data()
> #11 /var/www/html/wp-includes/script-loader.php(3445): {closure}()
> #12 /var/www/html/wp-includes/class-wp-hook.php(310): {closure}()
> #13 /var/www/html/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters()
> #14 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action()
> #15 /var/www/html/wp-settings.php(654): do_action()
> #16 /var/www/html/wp-config.php(124): require_once('/var/www/redacted...')
> #17 /var/www/html/wp-load.php(50): require_once('/var/www/redacted...')
> #18 /var/www/html/wp-blog-header.php(13): require_once('/var/www/redacted...')
> #19 /var/www/html/index.php(17): require('/var/www/redacted...')
> #20 {main}
> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(2161): Redis->set()
> #1 /var/www/html/wp-content/object-cache.php(294): WP_Object_Cache->set()
> #2 /var/www/html/wp-content/plugins/redacted/redacted.php(3106): wp_cache_set()
> #3 /var/www/html/wp-includes/class-wp-hook.php(310): wordpress_maintenance_class::pre_option()
> #4 /var/www/html/wp-includes/plugin.php(205): WP_Hook->apply_filters()
> #5 /var/www/html/wp-includes/option.php(150): apply_filters()
> #6 /var/www/html/wp-includes/theme.php(317): get_option()
> #7 /var/www/html/wp-includes/theme.php(328): get_template()
> #8 /var/www/html/wp-content/themes/twentytwentytwo/functions.php(67): get_template_directory()
> #9 /var/www/html/wp-settings.php(600): include('/var/www/redacted...')
> #10 /var/www/html/wp-config.php(124): require_once('/var/www/redacted...')
> #11 /var/www/html/wp-load.php(50): require_once('/var/www/redacted...')
> #12 /var/www/html/wp-blog-header.php(13): require_once('/var/www/redacted...')
> #13 /var/www/html/index.php(17): require('/var/www/redacted...')
> #14 {main}
> [19-Oct-2023 08:52:08 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2159
> Stack trace:
> #0 /var/www/html/wp-content/object-cache.php(2159): Redis->setex()
> #1 /var/www/html/wp-content/object-cache.php(294): WP_Object_Cache->set()
> #2 /var/www/html/wp-includes/option.php(990): wp_cache_set()
> #3 /var/www/html/wp-content/plugins/redis-cache/includes/class-plugin.php(1058): set_transient()
> #4 /var/www/html/wp-includes/class-wp-hook.php(310): Rhubarb\RedisCache\Plugin->do_admin_actions()
> #5 /var/www/html/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters()
> #6 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action()
> #7 /var/www/html/wp-admin/admin.php(237): do_action()
> #8 /var/www/html/wp-admin/options-general.php(10): require_once('/var/www/redacted...')
> #9 {main}

Reply:
>Hi again, OK, we now have the log data showing multiple errors when disabling Redis ... (please note that the errors persist after the disable Redis cache flush has finished running ...)

Would you mind explaining that in detail?

Reply:
OK. So, I've added `error_log()` to establish when flushing is starting and when it ends (as you say Redis cache flushing is blocking) ...

```
                    if ( $action === 'disable-cache' ) {
                        $result = $wp_filesystem->delete( WP_CONTENT_DIR . '/object-cache.php' );

                        if ( $result ) {
error_log('Before disable cache flush  ...');
                            try {
                                $predis = new Predis();
                                $predis->flush();
                            } catch ( Exception $exception ) {
                                // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
                                error_log( $exception );
                            }
error_log('... after disable cache flush');
                        }
```

With that in mind, when disabling Redis with the option in your plugin settings, this is what happens in the log ...

This shows the cache flush is about to start ...

> [19-Oct-2023 08:52:06 UTC] Before disable cache flush ...

Then we get two timeouts ... these could be a race condition whereby other requests are trying to talk to the cache whilst it's flushing or it could be the cache flush command itself, you would probably have more of an insight ...

> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161
> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161

This shows the cache flush command has completed ...

> [19-Oct-2023 08:52:07 UTC] ... after disable cache flush

Then we get three more timeouts that suggests that the cache flush has in fact **not** completed seeing as requests to the cache are still not working ... perhaps because the flush command is not blocking or because the flush command has timed out and the flush is continuing nonetheless ...

> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161
> [19-Oct-2023 08:52:07 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2161
> [19-Oct-2023 08:52:08 UTC] RedisException: read error on connection to 127.0.0.1:6379 in /var/www/html/wp-content/object-cache.php:2159

Your suggestion here is to increase the timeout on the flush itself but I don't see how you can possibly choose a timeout setting because you can't possibly know how long the cache will take to flush.

This is why I think that the cache flush needs to be hooked onto a hook like `shutdown` so that it doesn't mess with the existing cache requests in the current page request.

Furthermore, I think that ... just like WordPress uses a `.maintenance` file to prevent access to the site whilst updates are being performed, I think that `object-cache.php` should be renamed (deleted perhaps) temporarily whilst a flush is in operation to prevent concurrent requests from failing due to the cache flushing.

If you cannot tell when a cache flush has finished, it may be that the PHP checks every few seconds by running a test cache request and when a successful cache request has been made the cache flush can be considered to have been completed and the `object-cache.php` file put back.

As an aside, we have had to disable Redis on our main site again because even with PHP Redis the page load over time (a week or so) becomes stagnant and eventually the site is significantly faster without Redis. This is a separate issue which I will investigate at some point.

Reply:
Thanks for the details. I guess a good next step would be to narrow down the `RedisException: read error on connection` exceptions and attach more details so we know which Redis operations fail, as well as which PHP process-id is causing it. I think you're correct with the racing.

>This is why I think that the cache flush needs to be hooked onto a hook like shutdown so that it doesn't mess with the existing cache requests in the current page request.

Since you're the only user experiencing/reporting this, it's too large of change for the other 100k users. I assume your dataset is rather large and I reckon switching to Object Cache Pro and using the asynchronous flushing is a simpler solution. If you want a heavily discounted license email me 👍

Reply:
Honestly, it's not worth it.

So far there are three sites of ours that _were_ running Redis ([Rye](https://ryesussex.uk), [Hofnar](https://hofnar.com) and [Rush Witt & Wilson](https://rushwittwilson.co.uk)) and **all three** perform better and are more reliable without it.

We were hoping to setup persistent object caching on [Hawkhurst Parish Council](https://hawkhurst-pc.gov.uk) website too but they're unlikely to go ahead now.

> How about this: **I’ll pay you US$5000 if you can prove that I am wrong and open a PR with the fix.**

By my calculation you [owe me about 1,500 coffees](https://webd.uk/product/support-us/) by the way.

Reply:
While my offer still stands, again, there is nothing "broken" in Redis Object Cache, it's a Redis Server timeout.

All Redis operations are atomic and a long flush will block other processes and cause timeouts.

1. Increase the `WP_REDIS_READ_TIMEOUT` and you won't run into the timeout
2. Alternatively use `lazyfree-lazy-user-flush` to make flushes instant (best solution IMO)
3. Your suggestion (this PR) of pushing the timeout into `shutdown` will not prevent the timeout from happening

The #491 PR introduces a dedicated `WP_REDIS_FLUSH_TIMEOUT`, so you can control command and flush timeouts separately. It might need a 2nd iteration.

Reply:
OK, so the problem here is that my [original support request](https://wordpress.org/support/topic/page-timeout-2/) has been diluted by the occurrence of the timeout errors which (as I had thought) were not to do with the issue I was experiencing but instead due to a completely different bug as outlined in this PR.

My original problem (which persists) is that in every occurrence of using your plugin with Redis, the performance of the website(s) degrade sometimes to such an extent that the page requests (not the Redis requests) time out.

The fix for this when it happens (usually after a week or so of Redis being active) is to delete your `object-cache.php` and then put it back again. This action requires no restart of Redis or flush of Redis' cache.

We can't be doing that all the time so we've decided to ditch Redis (and your plugin) altogether with a view to explore Memcached in due course but to be honest, all our sites that purport to require a persistent object cache run faster without one. So we'll probably stick with the status-quo and ignore the warning in WordPress Site Health.

Thanks for your input and I do sincerely hope you get to the bottom of it one day.